### PR TITLE
Add Actions for Windows Installer

### PR DESCRIPTION
**Pull request by tobylane**
**Describe what the proposed change does**
- Builds a Windows installer from successful x64 and x86 Windows Actions runs, chosen by their run id from the url.
 example https://github.com/tobylane/CorsixTH/actions/runs/9158967469 is made of two runs from March.

### CODE REVIEW

**Review by lewri — CHANGES_REQUESTED** [review]
Microsoft Visual C++ 2015 - 2022 Redistributable appear to be missing from this installer (msvc140 and vcruntime140)

image

It installs and launches though, up to that point

**Comment by TheCycoONE** [comment]
I copied the files from my VS install manually

**Comment by TheCycoONE** [comment]
https://github.com/GitCommons/cpp-redist-nsis/blob/main/installer.nsh or something like it with the right version might be useful

**Comment by lewri** [comment]
Apparently the way of dropping them in the install folder isn't recommended. Microsoft encourages bundling the redist exe by the looks of things to install it systemwide (if needed). https://learn.microsoft.com/en-us/cpp/windows/redistributing-visual-cpp-files?view=msvc-170

However, I'm more than happy to continue supporting drag and drop of the required dlls instead to save space. Upon checking our current Appveyor and Github Actions builds though we don't have those bundle in the redist dlls (I assume the x86/x64/ARM versions are all different in retrospect). I'm not sure if the installer could run a check for the redists itself and download the installer itself if missing (I think I've seen some installers do that before).

**Comment by TheCycoONE** [comment]
Some do, if it was a wix/msi installer Microsoft ships a merge module to integrate with it, as we use nsis that's not an option. Supposedly there are different runtime dlls for different locales but I haven't heard of ours not working.

**Comment by tobylane** [comment]
Didn't mean to test in this branch, excuse the runs.

https://github.com/tobylane/CorsixTH/actions/runs/9257678748 has an installer using cpp-redist-nsis. Does that work?

File: WindowsInstaller/cpp-redist.nsh [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+!include LogicLib.nsh
+
+Section "customInit"
+  Var /GLOBAL VCRedistDownload
+  Var /GLOBAL isInstalled
+  ${If} ${RunningX64}
+    ;check H-KEY registry
+    ReadRegDWORD $isInstalled HKLM "SOFTWARE\Wow6432Node\Microsoft\VisualStudio\14.0\VC\Runtimes\x64" "Installed"
+    StrCpy $VCRedistDownload "http://aka.ms/vs/17/release/vc_redist.x64.exe"
```
Comment (lewri) [inline WindowsInstaller/cpp-redist.nsh:18]: Use https?

File: WindowsInstaller/cpp-redist.nsh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  ${EndIf}
+
+
+  ${If} $isInstalled != "1"
+    MessageBox MB_YESNO "NOTE: This application requires$\r$\n\
+      'Microsoft Visual C++ Redistributable'$\r$\n\
+      to function properly.$\r$\n$\r$\n\
+      Download and install now?" /SD IDYES IDNO VSRedistInstalled
+
+    ;if no goto executed, install vcredist
```
Comment (lewri) [inline WindowsInstaller/cpp-redist.nsh:31]: Is this meant to say `else install vcredist`?

File: WindowsInstaller/cpp-redist.nsh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    StrCpy $VCRedistDownload "http://aka.ms/vs/17/release/vc_redist.x64.exe"
+  ${Else}
+    ReadRegDWORD $isInstalled HKLM "SOFTWARE\Microsoft\VisualStudio\14.0\VC\Runtimes\x86" "Installed"
+    StrCpy $VCRedistDownload "http://aka.ms/vs/17/release/vc_redist.x86.exe"
+  ${EndIf}
+
+
+  ${If} $isInstalled != "1"
+    MessageBox MB_YESNO "NOTE: This application requires$\r$\n\
+      'Microsoft Visual C++ Redistributable'$\r$\n\
```
Comment (lewri) [inline WindowsInstaller/cpp-redist.nsh:27]: 2015-2022?

**Review by lewri — COMMENTED** [review]
This needs a fallback operation for the vcredist installer failing to download.
Also I tried with a network connection and it still skipped the vcredist and carried on installing (leading to missing dll errors again)

File: WindowsInstaller/cpp-redist.nsh (hunk 1 above)
Comment (tobylane) [reply WindowsInstaller/cpp-redist.nsh:18]: There's a limitation/unreliability in the download command and https.

File: WindowsInstaller/cpp-redist.nsh (hunk 2 above)
Comment (tobylane) [reply WindowsInstaller/cpp-redist.nsh:31]: Kind of. The goto is at the end, a different flow to if else.

File: .github/workflows/Windows.yml [hunk 4]
```diff
@@ -28,10 +28,10 @@ jobs:
   Windows:
     runs-on: windows-2022
     env:
-      PRESET: ${{inputs.preset || 'win-x64-rel'}}
-      VCPKG_DEFAULT_TRIPLET: ${{inputs.preset == 'win-x86-rel' && 'x86-windows' || 'x64-windows-release'}}
+      PRESET: ${{inputs.preset || 'win-x64-static'}}
+      VCPKG_DEFAULT_TRIPLET: ${{inputs.preset == 'win-x86-static' && 'x86-windows' || 'x64-windows-staticease'}}
```
Comment (lewri) [inline .github/workflows/Windows.yml:32]:
```suggestion
      VCPKG_DEFAULT_TRIPLET: ${{inputs.preset == 'win-x86-static' && 'x86-windows' || 'x64-windows-static-release'}}
```
I assume you meant that?

File: CMakePresets.json [hunk 5]
```diff
@@ -47,6 +47,30 @@
...
+    {
+      "name": "win-x64-static",
+      "binaryDir": "${sourceDir}/build/win-x64-static",
+      "generator": "Visual Studio 17 2022",
+      "architecture": "x64",
+      "cacheVariables": {
+        "CMAKE_TOOLCHAIN_FILE": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake",
+        "VCPKG_TARGET_TRIPLET": "x64-windows-static",
+        "ENABLE_UNIT_TESTS": "ON",
+        "DCMAKE_MSVC_RUNTIME_LIBRARY": "MultiThreaded"
```
Comment (TheCycoONE) [inline CMakePresets.json:60]: D should not be there

File: WindowsInstaller/LanguageStrings.nsh [hunk 6]
```diff
@@ -27,6 +27,7 @@ LangString options_title      ${LANG_DANISH} "Yderligere indstillinger"
 LangString save_in_appdata    ${LANG_DANISH} "Gemme spil data og indstillinger i %APPDATA% (Anbefales)"
 LangString remove_saves       ${LANG_DANISH} "Vil du gerne beholde dine gemte spil?"
 LangString overwrite_install  ${LANG_DANISH} "Det ser ud til at du forsøger at installere over en eksisterende installation af CorsixTH. Eventuelle ændringer mellem filerne kan gå tabt. Ønsker du at fortsætte?"
+LangString cpp_redist         ${LANG_DANISH} "NOTE: This application requires$\r$\n'Microsoft Visual C++ Redistributable'$\r$\nto function properly.$\r$\n$\r$\nDownload and install now?"
```
Comment (TheCycoONE) [inline WindowsInstaller/LanguageStrings.nsh:30]: If we get /MT compiling properly we don't need the redistributable anymore.

File: CMakePresets.json (hunk 5 above)
Comment (TheCycoONE) [reply CMakePresets.json:60]: https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html

Comment (TheCycoONE) [reply CMakePresets.json:60]: You also need to bump CMake's minimum version to 3.15 to use this flag.

**Comment by tobylane** [comment]
This uses [InstallRequiredSystemLibraries](https://cmake.org/cmake/help/latest/module/InstallRequiredSystemLibraries.html) to collect the dlls.

Installer made from this: https://github.com/tobylane/CorsixTH/actions/runs/9921377052

**Comment by lewri** [comment]
The latest installer functions as intended for me

**Review by lewri — COMMENTED** [review]
Are we now sticking with CMake 3.14?

**Comment by tobylane** [comment]
> Are we now sticking with CMake 3.14?

Yes, this cmake function is 3.10+.

**Comment by TheCycoONE** [comment]
Perfect, thanks @tobylane !

**Pull request merged**